### PR TITLE
Improved Stark-Zeeman Splitting

### PR DESCRIPTION
**Pull request by gardner-chris**
I have added the effect of zeeman splitting to the calculation of the spectra in FIDASIM. An example below shows the result in a uniform 5T B-field on the spectrum on the half energy part of a 1keV beam(so 0.5keV). In magenta is the spectra calculated taking into account both the Stark and Zeeman effect. The black dashed line is the result of only considering the Stark effect.

image_2021-05-18_151455

I verified the accuracy of the wavelengths in fidasim for both cases by comparing the wavelengths of the peaks to values calculated in python for 5T, 0.5keV.

5T1keV half bes spectral lines

I also added the stokes parameters to the spectra outputs. I verified that they were calculated correctly by comparing the first stokes parameter, s0, to the intensity since they should be the same. An example half bes stokes parameters calculated using a uniform B=1T and e_inj=1keV is in the image below. I also note that s1 for the pi lines should be the negative of s0 as another check. Currently s2 and s3 are always zero but I left them in incase we ever need to add some effect that will somehow cause them to be nonzero in the future.

half beam emission 4T1KeV stokes parameters

For details on the physics and derivations I also wrote this document: https://www.overleaf.com/read/nbwgnpfhsswb

### CODE REVIEW

**Comment by heidbrin** [comment]
I saw lambda_0=656.1 nm (the deuterium value) in your writeup.  Does your code use the correct value for a hydrogen neutral?

**Comment by gardner-chris** [comment]
> I saw lambda_0=656.1 nm (the deuterium value) in your writeup. Does your code use the correct value for a hydrogen neutral?

I did not input a new value for the wavelength. The spectrum subroutine has "lambda0" as an input and I used that. I believe the values are set with "line_lambda0 = [ 656.28d0, 656.104d0, 656.045d0 ] " in the subroutine identify_transition.

File: src/fidasim.f90 [hunk 1]
```diff
@@ -5231,42 +5288,82 @@ subroutine write_spectra
...
     endif
 
     if(inputs%calc_dcx.ge.1) then
         spec%dcx = factor*spec%dcx
+        spec%dcxstokes = factor*spec%dcxstokes
         if (inputs%stark_components.eq.0) then
             dcx = sum(spec%dcx, dim=1)
+            do k = 1,4
+                dcxstokes(k,:,:,:) = sum(spec%dcxstokes(:,k,:,:,:), dim=1)
+            enddo
```
Comment (lstagner) [inline src/fidasim.f90:5313]: Could this be replaced by a single `dcxstokes = sum(spec%dcxstokes, dim=1)`

File: src/fidasim.f90 [hunk 2]
```diff
@@ -5231,42 +5288,82 @@ subroutine write_spectra
...
     endif
 
     if(inputs%calc_halo.ge.1) then
         spec%halo = factor*spec%halo
+        spec%halostokes = factor*spec%halostokes
         if (inputs%stark_components.eq.0) then
             halo = sum(spec%halo, dim=1)
+            do k = 1,4
+                halostokes(k,:,:,:) = sum(spec%halostokes(:,k,:,:,:), dim=1)
+            enddo
```
Comment (lstagner) [inline src/fidasim.f90:5345]: Same as above

File: src/fidasim.f90 [hunk 3]
```diff
@@ -5285,81 +5382,134 @@ subroutine write_spectra
 
     if(inputs%calc_cold.ge.1) then
         spec%cold = factor*spec%cold
+        spec%coldstokes = factor*spec%coldstokes
         if (inputs%stark_components.eq.0) then
             cold = sum(spec%cold, dim=1)
+            do k=1,4
+                coldstokes(k,:,:,:) = sum(spec%coldstokes(:,k,:,:,:), dim=1)
+            enddo
```
Comment (lstagner) [inline src/fidasim.f90:5390]: same

File: src/fidasim.f90 [hunk 4]
```diff
@@ -5285,81 +5382,134 @@ subroutine write_spectra
...
     if(inputs%calc_fida.ge.1) then
         spec%fida = factor*spec%fida
-        if (inputs%stark_components.eq.0) fida = sum(spec%fida, dim=1)
+        spec%fidastokes = factor*spec%fidastokes
+        if(inputs%stark_components.eq.0) then
+            fida = sum(spec%fida, dim=1)
+            do k = 1,4
+                fidastokes(k,:,:,:) = sum(spec%fidastokes(:,k,:,:,:), dim=1)
+            enddo
+        endif
```
Comment (lstagner) [inline src/fidasim.f90:5425]: You know what to do....

File: src/fidasim.f90 [hunk 5]
```diff
@@ -5285,81 +5382,134 @@ subroutine write_spectra
...
     endif
 
     if(inputs%calc_pfida.ge.1) then
         spec%pfida = factor*spec%pfida
-        if (inputs%stark_components.eq.0) pfida = sum(spec%pfida, dim=1)
+        spec%pfidastokes = factor*spec%pfidastokes
+        if (inputs%stark_components.eq.0) then
+            pfida = sum(spec%pfida, dim=1)
+            do k = 1,4
+                pfidastokes(k,:,:,:) = sum(spec%pfidastokes(:,k,:,:,:), dim=1)
```
Comment (lstagner) [inline src/fidasim.f90:5487]: single sum

File: src/fidasim.f90 [hunk 6]
```diff
@@ -9400,53 +9601,99 @@ subroutine spectrum(vecp, vi, fields, lambda0, sigma_pi, photons, dlength, lambd
...
     endwhere
-
     !! normalize and multiply with photon density from colrad
+    normfactor = (1/sum(intensity))*photons*dlength
     intensity = intensity/sum(intensity)*photons*dlength
-
+    stokes(:,1) = stokes(:,1)*normfactor
+    stokes(:,2) = stokes(:,2)*normfactor
+    stokes(:,3) = stokes(:,3)*normfactor
+    stokes(:,4) = stokes(:,4)*normfactor
```
Comment (lstagner) [inline src/fidasim.f90:9693]: This can be replaced with a single `stokes = normfactor*stokes`

File: src/fidasim.f90 [hunk 7]
```diff
@@ -9501,18 +9751,21 @@ subroutine store_photons(pos, vi, lambda0, photons, spectra, passive)
...
             if (bin.lt.1) cycle loop_over_stark
             if (bin.gt.inputs%nlambda) cycle loop_over_stark
-            !$OMP ATOMIC UPDATE
+            !$OMP CRITICAL(spec_stokes)
             spectra(i,bin,ichan) = spectra(i,bin,ichan) + intensity(i)
-            !$OMP END ATOMIC
+            stokevec(i,1,bin,ichan) = stokevec(i,1,bin,ichan) + stokes(i,1)
+            stokevec(i,2,bin,ichan) = stokevec(i,2,bin,ichan) + stokes(i,2)
+            stokevec(i,3,bin,ichan) = stokevec(i,3,bin,ichan) + stokes(i,3)
+            stokevec(i,4,bin,ichan) = stokevec(i,4,bin,ichan) + stokes(i,4)
```
Comment (lstagner) [inline src/fidasim.f90:9765]: `stokesvec(i,:,bin,ichan) = stokesvec(i,:,bin,ichan) + stokes(i,:)`

**Comment by lstagner** [comment]
I noticed your tests have a lot of repeated code. It would be a good idea to consolidate the tests into a single module. Also a little documentation for the tests is warranted.

**Comment by lstagner** [comment]
Any updates when I can merge this in?

**Comment by gardner-chris** [comment]
> Any updates when I can merge this in?

Right now I am working on some corrections to the Lyman-Alpha stokes parameters and I still need to replace the do loops that you commented on. I'm nearly done with the corrections so it should be ready by the end of the week at the latest.

File: src/fidasim.f90 [hunk 8]
```diff
@@ -9640,8 +9627,12 @@ subroutine spectrum(vecp, vi, fields, lambda0, sigma_pi, photons, dlength, lambd
...
     q1 = sqrt(4*(e0*h*B/(4*pi*m))**2 + 9*(3*a_0*e0*E)**2)
     ! szratio is gamma/epsilon factor. can be thought of as roughly stark energy shift/ zeeman energy shift.
-    szratio = (e0*h*B/(4*pi*m))/(3*a_0*e0*E)
-
+    if(E.eq.0.d0)then
+    	szratio = 0.d0
+    else
+    	szratio = (e0*h*B/(4*pi*m))/(3*a_0*e0*E)
+    endif
+    allocate(circularity(n_stark))
```
Comment (lstagner) [inline src/fidasim.f90:9635]: Any reason your allocating `circularity` instead of just declaring it like you did earlier?

Comment (gardner-chris) [reply src/fidasim.f90:9635]: yes, having the array of length 3 for the Lyman-Alpha lines in the else if statement but an array of length 15 in the if statement makes it not work the way I had it before.

Comment (lstagner) [reply src/fidasim.f90:9635]: I'm confused. when you declare `circularity` as `real(Float64), dimension(n_stark) :: circularity` it should be equivalent to `allocate(circularity(n_stark))`

File: src/fidasim.f90 [hunk 9]
```diff
@@ -9503,11 +9503,11 @@ subroutine doppler_stark(vecp, vi, fields, lambda0, lambda)
         !+ Reference wavelength [nm]
     real(Float64), dimension(n_stark), intent(out) :: lambda
         !+ Wavelengths [nm]
-    real(Float64), dimension(15) :: wavel
```
Comment (gardner-chris) [inline src/fidasim.f90]: Oh setting 15 instead of n_stark must be the issue, not the way the array I was setting up the arrays. Will change in next commit

**Comment by gardner-chris** [comment]
By the Way, I think this is ready to merge unless you have more suggestions.

**Comment by lstagner** [comment]
The Travis CI is failing. Checkout the errors and see if you can fix them.
https://travis-ci.org/github/D3DEnergetic/FIDASIM/builds/772793773

**Comment by lstagner** [comment]
All right everything is passing now. I will merge this tomorrow unless you have anything further to add.

**Comment by alvin-garcia** [comment]
Attached is the report by @gardner-chris for this PR.
[stark_zeeman.pdf](https://github.com/D3DEnergetic/FIDASIM/files/6665042/stark_zeeman.pdf)

**Pull request merged**